v3/app/controllers/api/v3/user_controller.rb

```diff
@@ -0,0 +1,60 @@
+class Api::V3::UserController < ApplicationController
+
+  def sign_in
+    login = params[:login].to_s.strip.downcase
+    password = params[:password]
+
+    user = User.find_by(login: login)
+    if !user || !user.authenticate(password)
+      render :json => { :status => 400, :error => "Invalid." }.to_json and return
+    end
+
+    ip = request.remote_ip
+    timenow = Time.now.utc
+
+    # CREATE A TOKEN
+    token = UserToken.new_token(ip)
+    render :json => { :status => 400, :error => "Invalid." }.to_json if !token
+
+    user_token = UserToken.new
+    user_token.user_id = user.id
+    user_token.token = token
+    user_token.ip = ip
+    user_token.timenow = timenow
+    user_token.save
+
+    render :json => { :status => 200, :data => { :token => token } }.to_json
+  end
+
+  def sign_out
+    ip = request.remote_ip
+    token = params[:token].to_s.strip.downcase
+    all = params[:all] == "true" ? true : false
+
+    signout = User.sign_out({:ip => ip, :token => token, :all => all})
+    if !signout
+      render :json => { :status => 400, :error => "Invalid." }.to_json and return
+    end
+
+    render :json => { :status => 200, :data => { :message => "Signed out." } }.to_json
+  end
+
+  def validate_token()
+    ip = request.remote_ip
+    token = params[:token].to_s.strip.downcase
+    role_id = params[:role_id] ? params[:role_id].to_i : false
+
+    status, user = User.validate_token({:ip => ip, :token => token},role_id)
```

Comment:
Yes, I should move the token checking to a :before_action.

Reply:
DONE. Did NOT make it a before_action because we need to be able to check against a custom permission_id that is gated by the application itself and that is NOT a parameter in the API call itself. (e.g., the user needs admin permissions to make certain calls;  needs developer permissions to make other calls; etc.).  It is much easier to just call the method directly (and pass the custom parameter) than to try to automate that in a before_action.